docs/architecture/adr-002-api-cleanup.md

```diff
@@ -0,0 +1,94 @@
+# ADR ADR-002: API Cleanup to Improve Commit Performance
+
+## Changelog
+
+* 2023-06-06: First draft
+
+## Status
+
+DRAFT
+
+## Abstract
+
+This ADR proposes a cleanup of the API to make more understandable and maintainable the codebase of `iavl`.
+
+There is a lot of legacy code in the SDK that is not used anymore and can be removed. See the [Discussion](https://github.com/cosmos/iavl/issues/737) for more details. 
+
+There are some proposals for the speedup of the `Commit` by the async writes. See the [Discussion](https://github.com/cosmos/cosmos-sdk/issues/16173) for more details.
+
+## Context
+
+The current implementation of `iavl` suffers from performance issues due to synchronous writes during the `Commit` process. To address this, the proposed changes aim to finalize the current version in memory, parallelize hash calculations in `WorkingHash`, and introduce asynchronous writes in the background.
+
+It is also necessary to refactor the `Set` function to accept a batch object for the atomic commitments. 
+
+Moreover, the existing architecture of `iavl` lacks modularity and code organization:
+
+- The boundary between `ImmutableTree` and `MutableTree` is not clear.
+- There are many public methods in `nodeDB`, making it less structured.
+
+## Decision
+
+To make the codebase more clear and improve `Commit` performance, we propose the following changes:
+
+### Batch Set
+
+- Refactor the `Set` function to accept a batch object.
+- Eliminate the usage of `dbm.Batch` in `nodeDB`.
+
+### Async Commit
+
+- Parallelize hash calculations in `WorkingHash`.
+- Finalize the current version in memory during `Commit`.
+- Perform async writes in the background.
+
+### API Cleanup
+
+- Make `nodeDB` methods private and remove all unused methods.
+- Follow the naming convention of the `store` module.
+
+The modified API will have the following structure:
+
+```go
+    MutableTree interface {
```

Comment:
these are internal interfaces right? if these are public then i dont think this is fixing the current issue. The issue raised by dev is about the application only interacting with a single interface and making mutable and immutable tree internal workings of IAVL. 

what do you think of only having a reader and a writer interface.

Reply:
yeah, it is an internal separation, we will have one interface externally.
do you mean rename the mutable and immutable to reader and writer? or have external interfaces?